[doc] javascript for of 반복문 내용 추가 (#114)
joshua1988/vue-camp: docs/js/loop.md

hi
[]
```

## for of

> for of 반복문은 ES6에 추가된 구문으로서, 순회 가능한(iterable) 속성을 가지는 컬렉션에 사용하기 좋은 반복문입니다.

Reviewer: ```
:::tip
순회 가능한 객체란 Symbol.iterator 속성을 가지고 있고, iterator 객체를 반환하는 객체를 말합니다.
:::
```
148라인에 있는 설명을 이부분으로 옮기고, vuepress tip으로 보여주면 어떨까요?

Reviewer 2: iterable 속성에 대한 정의, 설명이 추가되면 좋겠습니다.

Author: 또한 iterable 속성에 대한 정의를 수정해봤습니다!
찾으면 찾을 수록 어려운 개념으로 빠지는 것 같아서 최대한 깔끔하게 정의해보려고 했으나, 
더 좋은 방식이 있다면 알려주시면 감사하겠습니다.

> for in 반복문과 비교하면서 알아보겠습니다. 아래와 같은 코드가 있다고 합시다.

Reviewer 2: 제안 -> `아래의 예시코드를 통해 for in 반복문과 비교하면서 알아보겠습니다.`

```js

> var arr= [10,20,30]
> let iterable="vue";

Reviewer: ```suggestion
var arr = [10,20,30]
let iterable = "vue";
```
컨벤션을 통일하면 깔끔해보이겠네요😄

```

위 배열과 문자열을 각각 for of와 for in 반복문으로 순회하면 아래와 같습니다.

```js
// 1) 배열을 순회하는 경우

> for(var num of arr){

Reviewer: ```suggestion
for (var num of arr) {
```
전체적인 for문에도 띄어쓰기를 추가해야 컨벤션에 맞겠네요

    console.log(num); 
}
for(var num in arr){
    console.log(num); 
}

// 2) 문자열을 순회하는 경우
for(let value of iterable){
    console.log(value); 
}
for(let value in iterable){
    console.log(value); 
}
```

각 코드의 결과는 아래와 같습니다.

```js
// 1) 배열을 순회하는 경우
// for of 반복문의 콘솔
10 20 30 

// for in 반복문의 콘솔
0 1 2

// 2) 문자열을 순회하는 경우
// for of 반복문의 콘솔
v u e

// for in 반복문의 콘솔
0 1 2
```

> 코드를 통해 알 수 있듯이, for in이 배열의 인덱스에 접근하는 데에 반해, for of은 배열의 값 자체에 접근하는 걸 알 수 있습니다.

Reviewer: ```suggestion
앞서 살펴본 코드를 통해 for in이 배열의 인덱스에 접근하는 반면 for of는 배열의 값 자체에 접근하는 것을 알 수 있습니다.
```
한 문장에 쉼표가 2개가 있어서 조금 다듬어보았어요

Reviewer 2: `코드를 통해 알 수 있듯이, for in이 배열의 인덱스에 접근하는 데에 반해, for of은 배열의 값 자체에 접근하는 걸 알 수 있습니다.`

--> for in 은 열거가능한 객체를 순회합니다. 배열도 객체이기 때문에 for in 으로 순회할 수 있으며, key에 해당하는 index를 반환하게됩니다.

이 내용이 언급되면 좋겠습니다.

Author: 새롭게 변경한 파일 부분에 for in의 접근방식에 대한 언급을 추가해봤습니다.
제가 커밋을 통해 수정한 방식이 옳은 것인지는 모르겠으나, 혹시 수정사항 확인이 어려우시면 다시 말씀 해주세요!

for in과 for of의 차이점은 아래와 같습니다.

## for in VS for of

- for in : 객체의 열거 가능한 속성에 대해 반복합니다.
- for of : 순회 가능한 속성을 가지는 컬렉션에 대해 반복합니다.
(* 순회가능한 것이란? : Symbol.iterator 속성을 가지고 있고, iterator 객체를 반환하는 객체를 말합니다.)

> 따라서 순회가능하지 않은 객체에 대한 for of 반복문 사용은 아래와 같이 에러를 일으킬 수 있습니다.

Reviewer: ```suggestion
따라서 순회 가능하지 않은 객체에 대한 for of 반복문 사용은 아래와 같이 에러를 일으킬 수 있습니다.
```

```js

> var obj={

Reviewer: ```suggestion
var obj = {
```

    num: 10,
    str: 'hi',
    arr: [],
}
for(var prop of obj){
    console.log(prop, obj[prop]); 
}
```

코드의 결과는 아래와 같습니다.

```js
Uncaught TypeError: obj is not iterable
```

> 순회가능하지 않은 객체에 대한 반복문 접근 에러에 대해서는 [이 자료](https://developer.mozilla.org/ko/docs/Web/JavaScript/Reference/Errors/is_not_iterable)를 참고하여 해결할 수 있습니다. 

Reviewer: ```suggestion
순회 가능하지 않은 객체에 대한 반복문 접근 에러에 대해서는 [이 자료](https://developer.mozilla.org/ko/docs/Web/JavaScript/Reference/Errors/is_not_iterable)를 참고하여 해결할 수 있습니다. 
```
